src/CVQualifiers.cpp

```diff
@@ -298,45 +320,37 @@ CVQualifiers::random_qualifiers(const Type* t, Effect::Access access, const CGCo
 	const Effect &effect_context = cg_context.get_effect_context();
 
 	// set random volatile/const properties for each level of indirection for pointers
+	// First set up the vectors with correct number of qualifiers:
+	unsigned level = 0;
 	const Type* tmp = t->ptr_type;
 	while (tmp) {
-		DEPTH_GUARD_BY_DEPTH_RETURN(2, ret_qfer);
-		isVolatile = rnd_flipcoin(volatile_prob);
-		ERROR_GUARD(ret_qfer);
+		++level;
+		is_consts.push_back(false);
+		is_volatiles.push_back(false);
+		tmp = tmp->ptr_type;
+	}
+	// Then make the random properties (properties need to be in reverse order): 
+	tmp = t->ptr_type;
+	while (tmp) {   
+		bool volatile_ok = is_volatile_ok_on_one_level(tmp);
+		isVolatile = volatile_ok? rnd_flipcoin(volatile_prob): false;
 		isConst = rnd_flipcoin(const_prob);
-		ERROR_GUARD(ret_qfer);
 		if (isVolatile && isConst && !CGOptions::allow_const_volatile()) {
 			isConst = false;
 		}
-		is_consts.push_back(isConst);
-		is_volatiles.push_back(isVolatile);
+		assert(level > 0);
+		is_consts[level-1] = isConst;
+		is_volatiles[level-1] = isVolatile;
+		--level;
```

Comment:
Hmm, seems we are modifying the semantics of the original code here. Seems we need to start from ```0``` instead of ```level-1```? E.g., 

    int level_2 = 0;
    while (tmp) {
        ...
        assert(level_2 < level)
        is_consts[level_2] = isConst;
        is_volatiles[level_2] = isVolatile;
        level_2++;
        ...
    }

Or, perhaps I am mistaken?